### PR TITLE
Manually defined CORS fitting for bagpipes

### PR DESCRIPTION
**Pull request by nightsh**
So, this was an adventure. Will make some notes for the future.

First, the `cors` module indirectly supplied by one of swagger's indirect dependency is no good here. Whatever you define in the configuration, will be partially overwritten by some magic internal libraries. Time to dig in, but before that I'll leave the configuration I tried here:

```js
  cors: {
    credentials: true,
    origin: [
      'https://tenders.exposed',
      'http://develop.tenders.exposed',
      'https://elvis-ember-develop.herokuapp.com',
      'https://elvis-ember-new-backend.herokuapp.com',
    ],
    // origin: true,  // We can use this if we want to automatically reflect request's origin
    methods: ['GET', 'PUT', 'POST', 'DELETE', 'OPTIONS', 'PATCH'],
    allowedHeaders: [ 'Origin', 'X-Requested-With', 'Content-type', 'Accept', 'X-Refresh-Token', 'Authorization'],
    optionsSuccessStatus: 200,
    preflightContinue: true,
  },
```

Second, I tried to use the [`swagger-express-middleware`](https://github.com/BigstickCarpet/swagger-express-middleware) library to inject a custom made middleware into swagger's own pipeline. This failed miserably, mostly because of the way we are creating the Swagger instance in `server.js` and my inability to refactor that by this library's taste.

Third, I had to `grep` around for whoever sets the `Access-Control-Allow-Origin` header. Found a promising match inside a default "fitting" of the `bagpipes` library. This seems to escalate the pipeline concept and bring it into Swagger by using some specifically tailored concepts, such as "pipes" and "fittings". After looking into it, I found a debug flag I could use to inspect the fittings and pipes flow, where I realised the following:

```
  pipes create fitting {"name":"cors"} +0ms
  pipes:fittings no user fitting cors in /home/victor/Workbench/elvis/elvis-backend-node/api/fittings +0ms
  pipes:fittings cors config: {"name":"cors"} +15ms
  pipes:fittings loaded user fitting cors from /home/victor/Workbench/elvis/elvis-backend-node/node_modules/swagger-node-runner/fittings +5ms
```

This part was by far the easiest, I created a fitting named "cors" in the designated location and moved the logic from `server.js` into it, with some sugar on top. Now we have control over CORS headers, however we want, since `bagpipes` won't interfere with our custom CORS fitting.

In a nutshell, quite a crazy puzzle for applying some goddamn headers.

### CODE REVIEW

File: api/fittings/cors.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+module.exports = function create() {
+  return (context, next) => {
+    const req = context.request;
+    const res = context.response;
+
+    const origin = req.headers.origin || '*';
```
Comment (georgiana-b) [inline api/fittings/cors.js]: @nightsh Good job! Sure looks cleaner :sunglasses: 
But wouldn't using `*` here cause the same troubles we have now that led to #81?

From what I understand clients will have to provide an `Origin` header. Don't you think it would be better to remove  default to `*` here and inform clients to supply `Origin` by [defining it in `swagger.yml`](https://github.com/tenders-exposed/elvis-backend-node/blob/master/api/swagger/swagger.yaml#L277-L281)?

**Comment by georgiana-b** [comment]
Btw @nightsh, if #81 is superfluous feel free to close.

**Comment by nightsh** [comment]
Here's [what the docs say](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) about this header. I guess all clients willing to log in with 3rd party OAuth will have to supply one, but that would be mainly browsers which pretty much fit the bill.

However, failing to supply a `Origin` header would result in an `undefined` value for preflight's allowed origin. In this case, a fallback of `'*'` was added, to prevent such behaviour and unintentional side effects in already existing implementations (this was the current default behaviour).

Issue #81 should be fixed by this PR as well. Will test and close if needed upon merging, as a separate, atomic task. Let's not mix them up – this PR adds possibility to manipulate CORS headers, if it doesn't fix #81 as well, then we manipulate some more so it would fix it. But they are different problems to solve.

To sum it up, this PR attempts to seamlessly patch the API, so what worked before still works and what wasn't because of wildcard allowed origin will start working.

**Comment by georgiana-b** [comment]
I just tested this rebased with the changes from `develop` and it does seem to solve the troubles described in #81, at least for Twitter.

Thank you for the contribution @nightsh!
Pls open future PRs against `develop` because sometimes `master` doesn't contain the latest commits. For eg in this case I pushed some changes in `develop` and was waiting for you to add the cors before merging in `master`.
